gh-94673: Add Per-Interpreter tp_subclasses for Static Builtin Types

ericsnowcurrently opened the pull request:
<!-- gh-issue-number: gh-94673 -->
* Issue: gh-94673
<!-- /gh-issue-number -->

ericsnowcurrently committed:
Add a note about tp_cache.

ericsnowcurrently committed:
Hide tp_subclasses behind functions.

ericsnowcurrently committed:
Distinquish between static and heap types.

ericsnowcurrently committed:
Move tp_subclasses for static types to PyInterpreterState.

ericsnowcurrently committed:
Add a note about heap types and tp_subclasses.

ericsnowcurrently committed:
Use PyDict_GetItemWithError().

ericsnowcurrently committed:
Add lookup___hash__().

ericsnowcurrently committed:
Handle the case of unhashable classes.

ericsnowcurrently committed:
Really handle the case of unhashable classes.

ericsnowcurrently committed:
Stop special-casing heap types for tp_subclasses.

ericsnowcurrently committed:
Revert "Stop special-casing heap types for tp_subclasses."

This reverts commit 0046b57446a26114fdf33d11a31a0fc84e7a764a.

ericsnowcurrently committed:
Point to the relevant GH issue.

ericsnowcurrently committed:
Only store builtins state on the interpreter state.

ericsnowcurrently committed:
state.subclasses -> state.tp_subclasses

ericsnowcurrently committed:
Do not set type.tp_subclasses for static builtin types.

ericsnowcurrently committed:
Store the index in tp_subclasses.

ericsnowcurrently committed:
Fix test_sys.

kumaraditya303 commented on Objects/structseq.c:
```diff
@@ -580,7 +580,8 @@ _PyStructSequence_FiniType(PyTypeObject *type)
     // Cannot delete a type if it still has subclasses
-    if (type->tp_subclasses != NULL) {
+    if (_PyType_HasSubclasses(type)) {
+        PyErr_Clear();
```
Where this exception is raised?

ericsnowcurrently replied:
Good catch.  `lookup_subclasses()` could fail before.  Now it can't.  I forgot to update all the call sites.  Now I have.

kumaraditya303 commented on Objects/typeobject.c:
```diff
@@ -4398,6 +4419,30 @@ type_dealloc(PyTypeObject *type)
+        return 0;
+    }
+    if (lookup_subclasses(self) == NULL) {
+        return PyErr_Occurred() ? -1 : 0;
```
same question

kumaraditya303 reviewed (approved):
LGTM, I have left some questions. 

FWIW, we could also get rid of the new type flag and use `tp_subclasses` as a tagged pointer and steal one bit to check if the `tp_subclasses` is per interpreter or local, I'll probably do that but for now this is fine.

kumaraditya303 commented on Objects/typeobject.c:
```diff
@@ -4300,10 +4307,14 @@ type_dealloc_common(PyTypeObject *type)
-    if (type->tp_subclasses == NULL) {
+    PyObject *subclasses = lookup_subclasses(type);
+    if (subclasses == NULL) {
+        PyErr_Clear();
```
same question

ericsnowcurrently committed:
Document the change.

ericsnowcurrently committed:
Make tp_subclasses void*.

ericsnowcurrently committed:
lookup_subclasses() never fails.

ericsnowcurrently committed:
Merge branch 'main' into shareable-static-types-tp_subclasses

encukou committed:
Adjust docs for ``tp_subclasses``

- Member documentation describes the current state, for users.
- The ``versionchanged`` note describes the change.
- The What's New entry is, ideally, usable for someone stuck with
  maintaining a mysterious legacy codebase that stopped working.

Details of internals are left out -- if it isn't enough to look
at the source, they should be documented in the devguide rather
than user-facing docs.

ericsnowcurrently committed:
lookup_subclasses() never fails.
